rali/rali/source/tf_record_reader.cpp

```diff
@@ -0,0 +1,301 @@
+        _in_batch_read_count = (_in_batch_read_count%_batch_count == 0) ? 0 : _in_batch_read_count;
+        std::string file_path = _folder_path;
+        file_path.append("/");
+        file_path.append(fname);
+        _last_file_name = file_path;
+        if(get_file_shard_id() != _shard_id )
+        {
+            incremenet_file_id();
+            file_contents.read(footer_crc, sizeof(data_crc));
+            continue;
+        }
+        _file_names.push_back(file_path);
+        incremenet_file_id();
+
+        _single_feature = feature.at("image/encoded");
+        _last_file_size  = _single_feature.bytes_list().value()[0].size();
+        _file_size.insert(std::pair<std::string, unsigned int>(_last_file_name, _last_file_size));
+	file_contents.read(footer_crc, sizeof(data_crc));
+        memcpy(&data_crc, footer_crc, sizeof(data_crc));
+        free(header_length);
+        free(header_crc);
+        free(footer_crc);
+        free(data);
+    }
+}
+
+void TFRecordReader::read_image(unsigned char* buff, std::string file_name, uint file_size)
+{ 
+    std::string temp = file_name.substr(0, file_name.find_last_of("\\/"));
+    const size_t last_slash_idx = file_name.find_last_of("\\/");
+    if (std::string::npos != last_slash_idx)
+    {
+        file_name.erase(0, last_slash_idx + 1);
+    }  
+    std::ifstream file_contents(temp.c_str(),std::ios::binary);
+    auto it = _image_record_starting.find(file_name);
+    if(_image_record_starting.end() == it)
+    {
+        THROW("ERROR: Given name not present in the map"+ file_name )
+    }
+    // std::cerr<<"\n image present at loc:: "<<it->second;
+    file_contents.seekg (it->second, std::ifstream::beg);
+    size_t uint64_size, uint32_size;
+    uint64_t data_length;
+    uint32_t length_crc, data_crc;
+    uint64_size = sizeof(uint64_t); 
+    uint32_size = sizeof(uint32_t); 
+    char * header_length = new char [uint64_size];
+    char * header_crc = new char [uint32_size];
+    char * footer_crc = new char [uint32_size];
+    file_contents.read(header_length, uint64_size);
+    file_contents.read(header_crc, uint32_size);
+    memcpy(&data_length, header_length, sizeof(data_length));
+    memcpy(&length_crc, header_crc, sizeof(length_crc));
+    char *data = new char[data_length];
+    file_contents.read(data,data_length);
+    _single_example.ParseFromArray(data,data_length);
+    _features = _single_example.features();
+    auto feature = _features.feature();
+    _single_feature = feature.at("image/filename");
+    std::string fname = _single_feature.bytes_list().value()[0];
+    if(fname == file_name)
+    {
+        _single_feature = feature.at("image/encoded");
+        memcpy(buff,_single_feature.bytes_list().value()[0].c_str(),_single_feature.bytes_list().value()[0].size());
+    }        
+    file_contents.read(footer_crc, sizeof(data_crc));
+    memcpy(&data_crc, footer_crc, sizeof(data_crc));
+    free(header_length);
+    free(header_crc);
+    free(footer_crc);
```

Comment:
Codacy found an issue: [Mismatching allocation and deallocation: footer_crc](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5763058)